components/datetime/src/date.rs

```diff
@@ -129,6 +129,157 @@ pub(crate) struct DateTimeInputWithLocale<'data, T: DateTimeInput> {
     calendar: Option<&'data week_of::CalendarInfo>,
 }
 
+/// A [`DateTimeInput`] type with all of the fields pre-extracted
+///
+/// See [`DateTimeInput`] for documentation on individual fields
+pub(crate) struct ExtractedDateTimeInput {
+    year: Option<Year>,
+    month: Option<Month>,
+    day_of_month: Option<DayOfMonth>,
+    iso_weekday: Option<IsoWeekday>,
+    day_of_year_info: Option<DayOfYearInfo>,
+    any_calendar_kind: Option<AnyCalendarKind>,
+    hour: Option<IsoHour>,
+    minute: Option<IsoMinute>,
+    second: Option<IsoSecond>,
+    nanosecond: Option<NanoSecond>,
+}
+
+/// A [`ZonedDateTimeInput`] type with all of the fields pre-extracted
+///
+/// See [`ZonedDateTimeInput`] for documentation on individual fields
+pub(crate) struct ExtractedZonedDateTimeInput {
+    date_time_input: ExtractedDateTimeInput,
+    gmt_offset: GmtOffset,
+    time_zone_id: Option<TimeZoneBcp47Id>,
+    metazone_id: Option<MetaZoneId>,
+    time_variant: Option<TinyStr8>,
+}
+
+impl ExtractedDateTimeInput {
+    /// Construct given an instance of a [`DateTimeInput`].
```

Comment:
Nit: Add a `TODO` to some backlog/1.x issue that we should consider optimizing this to only fetch the fields we need for formatting